### PR TITLE
Handling error response when a comma is present in the operand name of tags

### PR DESCRIPTION
**Pull request by shivam-sehgal**
### 🤔 What's changed?
to handle tags cases like @step1,@step2

adding regex to handle cases if the operand name of tags contains commas, in future this regex can be further extended
to handle any new case that will arise

### ⚡️ What's your motivation? 
heavily use this library, and wanted to contribute.
### 🏷️ What kind of change is this?
I came across issue [issue](https://github.com/cucumber/cucumber-jvm/issues/2776). Although the documentation specifies the expected format for tags, it seems that the pattern used by the user, '@step1,@step2,' did not generate any error. This discrepancy has the potential to confuse users. To address this, it would be beneficial to implement a validation mechanism that throws an appropriate error when such patterns are encountered.

- :bank: Refactoring/debt/DX (improvement to code design, tooling, documentation etc. without changing behaviour)
- :bug: Bug fix (non-breaking change which fixes a defect)
- :zap: New feature (non-breaking change which adds new behaviour)
- :boom: Breaking change (incompatible changes to the API)

### 📋 Checklist:

added support for regex filtering on operand names of tags
Limitations
1)Currently only handling case ',' in operand name
2)Currently just modified the java module of this lib

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

**Comment by jkronegg** [comment]
I like the idea of having cucumber notifying the user of misconfiguration issues. However, the proposed exception message (_Tag expression \"%s\" could not be parsed because of syntax error: Tag names should follow this pattern \"%s\"_") is a bit rough. 

As the natural understanding of ”@tag1,@tag2" is a an "or/and" combination of both tags, the exception message could be friendlier and suggest to use one of the operators (e.g. something like "It looks like your are trying to combine tags using comma (','), which is not supported as tag combination operators. Please replace it by either 'or' or 'and' operators, e.g. '@tag1 or @tag2'."). 

The PR missed unit tests.

File: java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java [hunk 1]
```diff
@@ -106,6 +108,7 @@ private static List<String> tokenize(String expr) {
                 isEscaped = true;
             } else if (c == '(' || c == ')' || Character.isWhitespace(c)) {
                 if (token.length() > 0) {
+                    isOperandValid(token,expr);
```
Comment (jenisys) [inline java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:111]: The name of the method `isOperandValid()` is actually wrong here.
Correct, that is your intention but also operator-strings like `and`, `but` and `not` are checked by this method.

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java]: Thank you, @jenisys  , for your review and feedback. I have made the necessary adjustment by updating the method name. As you rightly mentioned, apart from operands, this method will also handle validations for operators. Please confirm if the changes meet the requirements. Your continued assistance is appreciated

**Comment by shivam-sehgal** [comment]
> I like the idea of having cucumber notifying the user of misconfiguration issues. However, the proposed exception message (_Tag expression "%s" could not be parsed because of syntax error: Tag names should follow this pattern "%s"_") is a bit rough.
> 
> As the natural understanding of ”@tag1,@tag2" is a an "or/and" combination of both tags, the exception message could be friendlier and suggest to use one of the operators (e.g. something like "It looks like your are trying to combine tags using comma (','), which is not supported as tag combination operators. Please replace it by either 'or' or 'and' operators, e.g. '@tag1 or @tag2'.").
> 
> The PR missed unit tests.

Thank you, @jkronegg, for your valuable comment and thorough review. I have enhanced the error message with additional details to provide clearer guidance. Additionally, I have included relevant unit tests to ensure proper validation. Please review the changes and kindly confirm if they meet the requirements. Your continued support is greatly appreciated.

File: java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java (hunk 1 above)
Comment (jenisys) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java]: The method-name is now OK for me 👍

**Comment by jenisys** [comment]
HINTS:
* The change in the test data affects all implementations (Java, Go, …)
* The implementation is currently only provided for Java
* Therefore, the tests for all other languages that lack this implementation have test failures

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
This PR prevents the use of commas inside tag expressions altogether.

However comma's are valid characters inside a tag. Consequently they must be a valid character to use in a tag expression. For example a scenario may be tagged with `@a1,b2`. Which could then be selected by the tag expression `@a1,b2`

```feature
  @a1,b2
  Scenario: Example 1 
```

Likewise a scenario might be tagged with `@a1,` and `@b2`.  This should be selectable through the tag expression  `@a1, and @b2`.

```feature
  @a1,@b2
  Scenario: Example 2

  @a1, @b2
  Scenario: Example 3  
```

The intent of this PR is to inform users they are using a tag-expression syntax from v1 while using v2. Currently I do not believe there way to accept all valid v2+ tag expressions while rejecting all v1 expressions. There may however the possibility of validating literals. 

A tag in Gherkin is delimited by `@` and Gherkin does not allow the `@` to be escaped. So every tag matches the pattern `^@[^@]+`. So any tag expression literal that contains a `@` after the starting `@` can not possibly target a gherkin tag.

For example

```feature
  @a1@b2
  Scenario: Example 4
```

This scenario is tagged with `@a1` and `@b2` rather than `@a1@b2`. So the tag expression `@a1@b2` will never find any valid scenarios and could be considered invalid. 

This will incidentally also capture any use of the v1 tag expressions of the form `@a1,@b2` and also `~@a1`. 

Proposal:
 * Validate that Gherkin tags do indeed follow the `^@[^@]+` pattern
 * Validate literals in tag expressions to match the `^@[^@]+` pattern.
 * (Optional) Provide the user with examples of correct tag expressions in the error message. Note this should be done for all parse errors, not just the one's we are adding.

File: java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java [hunk 2]
```diff
@@ -11,7 +11,7 @@
 
 public final class TagExpressionParser {
     //regex for token to ensure no token has ',' in them later can be customized further
-    private final static String VALID_TOKEN = "(?:@[^@]*|and|or|not|/(|/))$";
+    private final static String VALID_TOKEN = "^(?:@[^@]*|and|or|not|\\(|\\))$";
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: By trying to validate all tokens rather than the litterals you are making this more complex than needed. Considering that this implementation will have to be repeated for a few languages, perhaps a review in the middle can save some duplication of effort.

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: Thanks, @mpkorstanje, for your reply
I assume that a tag expression consists of operands and operators. When splitting the expression, there are two possibilities. The current token or string could be an operator, which includes "(", ")", "and", "or", and "not". Alternatively, it could be an operand, which should adhere to the Gherkin standard of starting with "@" and not containing "@" anywhere else.

To handle this, I have used a regex pattern to ensure that only valid tokens, which form the expression, are evaluated. By considering both operands and operators, it simplifies the code and eliminates the need for additional checks to determine the token type.

I believed this approach reduces complexity and improves the readability of the code, but maybe I might be missing someting, I would appreciate your opinion on this.

Comment (jenisys) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: Well, actually the situation is slightly different (as you can see from the tests / testdata):

* Tags in a tag-expression may either be `@foo` (with-leading-atsign) or `foo` (without-leading-atsign)
* `@` (atsign) should not occur in a tag (after the first char) in a tag-expression
* Gherkin parser removes the `@` (atsign) from tags for Features/Rules/Scenarios/… . Therefore, `@foo` becomes`“foo“` internally

SEE ALSO:
* [testdata/parsing.yml](https://github.com/cucumber/tag-expressions/blob/main/testdata/parsing.yml)
* [testdata/errors.yml](https://github.com/cucumber/tag-expressions/blob/main/testdata/errors.yml)
* [python/tests/functional/test_tag_expression.py#L57](https://github.com/cucumber/tag-expressions/blob/main/python/tests/functional/test_tag_expression.py#L57)

Comment (mpkorstanje) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: > Gherkin parser removes the @ (atsign) 

@jenisys are you sure about that? Or is that true only for some Gherkin parsers?

Comment (mpkorstanje) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: >  but maybe I might be missing someting, I would appreciate your opinion on this.

@shivam-sehgal without repeating myself that will be a bit difficult.

But do consider validating only the litterals. In essence, do not put the validation in the middle of tokenizer but rather after tokenization, just prior to litteral creation.

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: Thank you, @jenisys, for your input. It makes sense to allow users to pass tags without the leading `@` symbol. My initial pull request was based on @mpkorstanje's [proposal](https://github.com/cucumber/tag-expressions/pull/107#pullrequestreview-1520893810), which followed the pattern `^@[^@]+`, which makes sense as per tags details in the Gherkin documentation. However, it's important to note that nowhere in the documentation does it explicitly state that tags should always follow this pattern.

Considering your advice to restrict the use of `@` after the first character and to allow tag names without the leading `@`, I propose two options for changing the regex in this pull request:

Option 1:
Regex: `/^(?:@[^@]*|[^@]*)$/`
This option allows two types of tokens in the tag expression:
1. Strings that start with `@` but do not have `@` anywhere after the first character.
2. Strings that do not contain the `@` symbol at all.

Option 2:
Regex: `/^(?:@[^@]*|.*)$/`
This option also allows two types of tokens in the tag expression:
1. Strings that start with `@` but do not have `@` anywhere after the first character.
2. All other strings without any restrictions.

I would appreciate it if both you and @mpkorstanje could review these suggestions and provide guidance on the final decision. Once we reach a consensus, I can proceed with the final code changes.

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: Thank you, @mpkorstanje, for your explanation. It makes sense and sounds better. I will work in the direction of applying the proposed `^@[^@]+` pattern specifically to literals, excluding operators. Will remove my checks from the tokenization process, and focus on applying the checks to the part where literals are created.

Comment (mpkorstanje) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: Don't go to quickly.  :)

The first step of the proposal was to validate that Gherkin tags do indeed follow the ^@[^@]+ pattern. It looks like @jenisys believes that is not the case.

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: agree @mpkorstanje, It appears that in the tags references in Gherkin documentation, they consistently follow the pattern you proposed with tags starting with `@`. Although the documentation does not explicitly mention that tags should follow this pattern, all the examples given in the documentation support this assumption. However, before proceeding, I would like to hear @jenisys' thoughts on this matter :)

Comment (jenisys) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: > Don't go to quickly. :)
> 
> The first step of the proposal was to validate that Gherkin tags do indeed follow the ^@[^@]+ pattern. It looks like @jenisys believes that is not the case.

* My case was only that the initial `isOperandValid()` method did not only check `Literal`(s)/Operand(s) but also operators.
* Actually, by stepping back: Operators and Groupings do not need to be checked because the parsers ensures that anyway (after the tokenization phase)

The primary intention for the change was to check that `Operand`(s) (aka: `tag`(s) ) should not contain `comma` inside to better detect user problems (primarily caused by tag-expression v1 (IMHO) where `comma` was a valid logical binary operator):

* Maybe we should perform this check at another place (line) were only operands occur (for example: Shortly before `Literal` objects are created).

OTHERWISE (as @mpkorstanje stated):

* There may be cases where `comma` is already used as part of a `tag`
* I partly agree with @mpkorstanje that these cases exist (example: Some `ActiveTag` in `behave` where a comma-separated list is used as part of the `value`-part, like: `@use.with_platforms=linux,macos`
* But the question for me is: would `tags-with-commas` be used within a `tag-expression` to select Feature(s)/Scenario(s)/… . My personal answer to this question is currently `no`, at least not for `ActiveTag`(s) (aka: `skip-if` with positive Boolean logic)

MY RESUMEE:

* Maybe the primary problem can be prevented by another solution

Comment (jenisys) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: A potential solution could also be:

* User can select/`opt-in` a checker that ensures that `tag`(s) do not contain `comma`(s)
* The everyday user does not need to use this check (especially if he/she uses `tag-with-comma`(s) )

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: > that

Thanks, @jenisys, for your valuable comments.

Regarding your point about token parsers accepting tags without a leading `@`, and the use of tags without a leading `@` in many unit test cases, it is true that unit tests are primarily focused on testing the token parser logic, which currently allows operands/tags without leading  `@`. However, when we consider the bigger picture and the use of tags in cucumber feature files, adding tags without a leading `@` cause an issue. The [FeatureParser](https://github.com/cucumber/cucumber-jvm/blob/main/cucumber-core/src/main/java/io/cucumber/core/feature/FeatureParser.java) throws an exception when it encounters a tag line without a leading `@`, as it is unable to recognize it as a valid tag. 

Therefore, while unit tests may deviate from the Gherkin standard pattern, it is important to align them with the pattern which I am making in this PR. @mpkorstanje's proposal of checking the `^@[^@]+` pattern for tags makes sense. It ensures that tags follow the expected format and adheres to the Gherkin standard. By following this regex pattern we can avoid confusion among users of using `@tag1,@tag2 ` as tag expression, which tag parser logic doesn't comply with, which is also the main purpose of this PR.

I would appreciate hearing both of your thoughts on this matter.

cc: @mpkorstanje

Comment (jenisys) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: @shivam-sehgal @mpkorstanje 
Mmh, `tag-expressions` and `tags` in Gherkin files are two different things. My points related only to `tag-expressions`. It is clear that the Gherkin parser needs `@some_tag` to detect a `tag`.

If you perform the check in the `TagExpressionParser` that `tags` in a `tag-expression` require `@` prefix:

* you need to change most `testdata` cases/examples
* you need to adapt all `tag-expresions` parser implementations
* you to need to inform all existing users of `tag-expressions` that you have a breaking-change and that they need to adapt their `tag-expressions` (or the package maintainers need to do that)

Personally, I have no problem with this change. The usage of `@foo or @bar` makes it easier for users to relate the `tag-expression` to `tags` in Gherkin files (even if that is not how the initial/existing implementation works).
But I wanted that you are aware of that (before anything is changed).

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: Thank you, @jenisys, for expressing your concerns in such a descriptive manner.

Regarding the issue of changing test data, I want to assure you that I will address that. The implementation part will definitely handle this effectively by using minimal, readable, and maintainable code.

A major concern I have noticed is the need to inform users. I understand and acknowledge this concern. When considering the tag-expression, it falls under the umbrella of Cucumber, which enforces [Gherkin](https://cucumber.io/docs/guides/overview/) grammar rules. Currently, the code deviates from the standard set by the framework this dependency is a part of. I strongly believe that enforcing this pattern will be beneficial for future development. It will not only make it easier to handle errors related to combinations of literal values passed by users but also ensure adherence to the standard (Gherkin) that Cucumber highly emphasizes.
@mpkorstanje 
Please let me know if you see any concern in this, it will be highly appreciated!

Comment (mpkorstanje) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: > I have not checked, all existing Gherkin parsers
> The ones that I know provide tags without leading @ prefix-char

@jenisys but which are the implementations you know? 

@shivam-sehgal since the assumption that literals must start with an `@` doesn't hold I don't think it is possible to inform users that they are using the v1 syntax in this library. Perhaps `cucumber-jvm` would be a better place to do so.

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: @mpkorstanje cucumber-jvm basically calls this library to parse the tag expression as well as to validate if the tag expression is valid or not so the flow is like this in cucmberOptions annotation user provides the tags expression shown in the below code
`@CucumberOptions(features = "src/it/resources/features",
        glue = {"com.adobe.aep.devex.exim.integrationtests.stepDefinitions",
                "com.adobe.aep.devex.exim.integrationtests.config",
                "com.adobe.aep.devex.exim.integrationtests.hooks"
        },
        plugin = {"pretty", "html:target/cucumber-html-report.html",
                "json:target/cucumber.json",
                "rerun:target/cucumber-api-rerun.txt"
        },
    tags = "not @ignore and @hello" // Exclude the feature with the "@ignore" tag
      )`
this annotation is parsed by [CucumberOptionsAnnotationParser.class](https://github.com/cucumber/cucufmber-jvm/blob/main/cucumber-core/src/main/java/io/cucumber/core/options/CucumberOptionsAnnotationParser.java).
addTags method in this class is called which then calls this repo to validate the tag expression, now the intent of this PR was users by mistake put tag expressions like @tag1,@tag2 in the tags section in the CucumberOptions, and they don't get any error message, to avoid that we need to validate the tag expression not having any tags or literals like `@tag,@againInTag` to prevent this confusion and give users a valid error response so they can understand that they need to use logical operator, instead of commas, I can put this tag validation for the provided tag expression in addTags method of the CucumberOptionsAnnotationParser clas i.e in cucumber jvm, but my mind says it's better to keep validation logic inside this tag expression validation repo only to follow the single responsibility principle of oops,
to address this concern with the least impacting change, for now, I see if there is an expression having a tag starting with `@` that tag shouldn't have `@` after first, it will prevent this confusion with a proper error message letting user understand that they are doing mistake and cucumber is reading their expression as a single tag, and also won't affect any case where we allow the literals that are not starting with `@`

In short, cucumber-jvm calls this repo to validate the tag expression passed by the user and uses a message from `TagExpressionException`to let the user know with a message what is wrong with the expression, the message should be coming from here i.e from this repo's java directory, else the responsibility of this code we have to handle separately in cucumber-jvm.

Comment (jenisys) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: One solution approach could be:

* if `CucumberOptions` had an optional parameter, like `tag_expression_validator`, `tag_expression_parser` or an `tag-expression preprocess hook` you could perform a pre-process check
* Pre-process check would just split the `tag-expression` string value into words and check if `comma`(s) are contained in any word
* If a `comma` is contained, you issue a log-warning or raise an exception (whatever the „right“ consequent action is), …

This approach would have the following advantages:

* Strict checking is optional
* If you want/need strict checking of `tag-expressions`, you can plug-in the functionality

HINT:
By looking at the current implementation of the `CucumberOptions` class, this needed `parameter` would need to be added to support such a kind of solution.

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: @jenisys, I find the approach of using an optional flag for literal validation quite appealing. However, I have some doubts about the term "strict checking." If libraries using this repository are enforcing Gherkins rules, then it seems reasonable for this library to do the same. However, if we believe and would like to continue supporting users who define literals not starting with `@` in new versions, then the optional parameter approach makes sense.

One potential concern raised by @mpkorstanje is that checking for commas alone may cause issues. To differentiate between v1 and v2, perhaps we should consider checking for `^@[^@]+` which he mentioned if the optional strict validation flag is passed in `CucumberOptions`.

I would appreciate hearing both of your opinions on this matter. Let's collaborate and I can begin working on it.

cc: @mpkorstanje

Comment (mpkorstanje) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: I think the cost of the complexity is starting to exceed the value of the solution. 

Perhaps a simpler solution such as improving the documentation in Cucumber JVM might help?

Comment (shivam-sehgal) [reply java/src/main/java/io/cucumber/tagexpressions/TagExpressionParser.java:14]: thanks @mpkorstanje for your advice, have opened a pull request to update the doc a little bit, to add info for users to avoid the confusion of separating tags using `,`, attaching the screenshot of the doc changes below
Added the screenshot of the change in the description of the [PR](https://github.com/cucumber/docs/pull/911), would like to request your review of this.
Thanks in advance
